### DISCUSSION TITLE
Potential Project: Crowdsourcing Corner Tracking Data

### DISCUSSION THREAD

**Discussion by larsmaurath** [DonsetPG/narya#27] [Ideas]
As you probably know, missing open access to great football event or tracking data is one of the main hurdles for many to start out in football analytics.

I think the Narya API opens the door to crowdsourcing some of this data. I think one sequence of play lends itself ideally for that: **corners**. One of the few standardized and static situations in football that is also little exploited (success from corners ranges widely between teams), but at the same time has a lot of potential (e.g. FC Midtjylland).

Please let me know if anybody is interested in working together as a small team to work on this. I see a few steps necessary that can easily be worked on in parallel.

1) Snapping video footage of corners
2) Generating training labels for both the homography and player tracking
3) Generating event labels
4) Possibly training a narya model solely on corner footage

A few issues that I see:

1) Given licensing issues with video footage this likely needs to be a small, closed group
2) Corners are likely the toughest job for a player tracking model (lots of overlap between players). Unsure how successful this will be from broadcast footage
3) Broadcast footage often times does not cover the full corner (zooms, unrelated frames). Hopefully we can get enough footage.

The ultimate goal would be to fully open-source this data for anybody to use. Ideally in some already existing format (e.g. Metrica for the tracking). 

Let me know any feedback!

**Comment by DonsetPG** [comment]
Hey,
This sounds like a great idea! 
A few thoughts: 

> Generating event labels

Just to be sure, by event labels you mean timestamp associated with an action like a pass or a kick? 

> Given licensing issues with video footage this likely needs to be a small, closed group

I used youtube videos. Wouldn't that be enough and skip this licensing issue? 

> Corners are likely the toughest job for a player tracking model (lots of overlap between players). Unsure how successful this will be from broadcast footage

I think this could be interesting, and it would even allow us to try other models such as body pose estimators (might be better when a lot of overlap is happening)  

> Ideally in some already existing format (e.g. Metrica for the tracking).

I agree. I think I made a mistake and should have try to follow their format, or at least build some kind of connector. 

__ 

Anyway, looks like a great project and you can count me in. I can try to spend some time building the footage and maybe some annotations. I can also try to test some models later on. 

Have a great day. 

@TheoGreg What do you think?

**Reply by larsmaurath** [reply]
Cool thanks for the feedback!

**On event labels:**
I think everything that would be interesting for subsequent research would be good to have, starting with the corner kick and subsequent 5 actions (or until shot/clearance)? If we manually do it I'd again replicate the format of something existing (Opta or Statsbomb?). Of course this shouldn't take that long per corner so nobody wants to do it. Maybe we can even reuse existing free data.

**On licensing**
Youtube should be fine, I guess I am maybe too careful on that one. I also have access too full CL matches of this season through my streaming platform. And there is always footballia.

I'll also start putting together some footage too see how long it takes me per game and how clean the footage is for a typical corner.

**Reply by DonsetPG** [reply]
Ok great!
Feel free to release maybe one example corner, so that others could try to replicate the labels on other footage.
Maybe we could even write a small readme regarding the labeling, to ensure consistency?

**Comment by alenm10** [comment]


**Reply by larsmaurath** [reply]
Great! I have used a little tool to generate homography, keypoint and player tracking labels for narya. I am currently cleaning it up and will then post the link here.

**Reply by larsmaurath** [reply]
I have now published the tool here: https://github.com/larsmaurath/narya-label-creator

**Comment by larsmaurath** [comment]
I put together an example of what I mean with video footage and event labels for the two Champions League matches on Wednesday. The footage seems good enough. I'd say on average 5% of corners are not shown at all, 10% have camera switches during the corner.

https://www.dropbox.com/s/dct3s1f9yuaj0n0/corner_examples.zip?dl=0

Few thoughts on short corners: an easy decision would be to simply ignore short corners, but the change in angle they aim to achieve would probably also be an interesting research topic. One approach could be to keep short corners that result in a cross within two/three touches.

FYI creating the footage took a bit more than a minute per corner after I had a good system.

**Reply by DonsetPG** [reply]
Hey I just went trough the footage+csv, it looks good. Did you have time to process more footages since then? 
Let me know when we can start helping you, I can try to find some time for player or key points labelling. 
Have a great weekend, cheers

**Reply by larsmaurath** [reply]
Haven't created more footage because I wanted to wait for feedback from everybody but looks like these are good examples.
I thought a nice goal for the project could be to collect data on all corners of this season's Champions League (definitely ambitious, but should give us a nice sample size across many different teams).

I think what would be most helpful at this point would be creating training labels and potentially training narya on corner labels only. I can provide the footage and event labels.

Do you have a good way to create the player tracking labels or have you already checked out (https://github.com/larsmaurath/narya-label-creator)? 

A few more open questions I came up with:
1) Once we have reliable player tracking we probably also want to split up the identified players into offensive/defensive team. I assume this should be relatively "easy" by identifying the dominant color in the bounding box or so.
2) Syncing up the event and tracking data. I think ideally we should know which frame coincides with the corner kick being taken. I think in the worst case this would need to be tagged again manually (maybe something for stage two).

**Reply by DonsetPG** [reply]
>  thought a nice goal for the project could be to collect data on all corners of this season's Champions League (definitely ambitious, but should give us a nice sample size across many different teams).

Haha why not, how many samples are we talking about in your opinion?

> Do you have a good way to create the player tracking labels or have you already checked out (https://github.com/larsmaurath/narya-label-creator)? 

We used RectLabel. I will try to work with your tool this weekend! 

> Once we have reliable player tracking we probably also want to split up the identified players into offensive/defensive team. I assume this should be relatively "easy" by identifying the dominant color in the bounding box or so. 

yes, the ReId model could also help with that. 

> Syncing up the event and tracking data. I think ideally we should know which frame coincides with the corner kick being taken. I think in the worst case this would need to be tagged again manually (maybe something for stage two). 

Unless I am mistaken, we should aim for:
-> Keypoints detections for homography: Seems like the model already works fine. Could be fine tuned tho. 
-> Player detection: a good amount of work to do. We can retrain a gluon model on a corner dataset, or see if new more robust methods exists. 
-> Event detection: I remember reading a few papers about this. Maybe something like a RCNN + player detections features could be enough? 

Overall I think a good target would be to: 
* Label events samples
* Label player detection samples 
* Read on new player detection methods 

Let me know what you think.

**Reply by larsmaurath** [reply]
> Haha why not, how many samples are we talking about in your opinion?

There are 125 games per season and with roughly 10 corners per game this would bring us to 1250. Of which may be 1000 are usable.

> Overall I think a good target would be to:
> 
>     Label events samples
>     Label player detection samples
>     Read on new player detection methods

Yes sounds good. Let me continue with creating the footage and I'll then share training images (1024 x 1024) in batches of 100. One image for every corner (I am currently at 76). Multiple people can then work on the labeling in parallel.

Hopefully we'll then see improvement in the player tracking when training the model solely on corner footage.

**Reply by larsmaurath** [reply]
Find below the link to the first batch. 177 frames from 100 corners for player and homography labeling.


https://www.dropbox.com/s/mlub6efhbj7wpfe/Batch_1_Frames.zip?dl=0

**Reply by DonsetPG** [reply]
Thanks a lot. I'll start with the players labeling this week-end!

**Reply by larsmaurath** [reply]
Cool! I'll continue to produce these batches for now. @alen1010 let me know if you are interested in labeling some of these frames (later batches). If not I am sure there will be other useful tasks later on.

**Reply by larsmaurath** [reply]
Second batch below: 191 frames from another 100 corners.

https://www.dropbox.com/s/u24px42toou4oog/Batch_2_Frames.zip?dl=0

**Reply by larsmaurath** [reply]
> If @DonsetPG is labeling players I can start homography labeling on the dataset you provided.

Sounds great. Do you have a good way to create the homography annotations? For now the labelling tool I referenced above only works for vertical pitches, but I am working on adapting it for horizontal ones/broadcast footage.

**Reply by larsmaurath** [reply]
Wanted to quickly check in if the below labelling looks good to everybody. Especially the handling of overlapping players. Below are screenshots from (https://github.com/larsmaurath/narya-label-creator).


<img width="510" alt="Screenshot 2021-04-02 at 13 26 56" src="https://user-images.githubusercontent.com/11316760/113439868-5a691100-93b9-11eb-9297-6361d70cbe49.png">
<img width="510" alt="Screenshot 2021-04-02 at 13 29 07" src="https://user-images.githubusercontent.com/11316760/113439870-5b01a780-93b9-11eb-9923-04694c017f09.png">
<img width="507" alt="Screenshot 2021-04-02 at 13 31 31" src="https://user-images.githubusercontent.com/11316760/113439871-5b9a3e00-93b9-11eb-8a9e-e4e2cbf466b6.png">

**Reply by DonsetPG** [reply]
Looks good to me, I am doing the same.

**Reply by larsmaurath** [reply]
> Looks good to me, I am doing the same.

Cool thanks!

**Reply by larsmaurath** [reply]
> > > If @DonsetPG is labeling players I can start homography labeling on the dataset you provided.
> > 
> > 
> > Sounds great. Do you have a good way to create the homography annotations? For now the labelling tool I referenced above only works for vertical pitches, but I am working on adapting it for horizontal ones/broadcast footage.
> 
> I started using this tool: https://kkoripl.github.io/NaryaKeyPointsDatasetCreator/
> Here are first 10 images annotated from the first batch you posted so let me know what you think: https://easyupload.io/ivb8t4

Of course; temporarily forgot about that great tool. Your examples look good to me! Cool, I think we are then set up to make progress in all directions!

**Reply by larsmaurath** [reply]
Probably a better question for @DonsetPG so we use the same approach that was used for training the initial narya release. When I trained narya on vertical pitches it worked well to mark all keypoints, view obstucted or not, but I'd refer to Paul.

**Reply by DonsetPG** [reply]
Yeah I think even obstructed points should be marked!

**Reply by larsmaurath** [reply]
Third batch below: 137 frames from another 100 corners.

https://www.dropbox.com/s/ah44g62x726tar3/Batch_3_Frames.zip?dl=0

**Reply by DonsetPG** [reply]
Thanks a lot. 
To be honest I have a lot of work at the moment so it's hard to find some time with the labelling. If anyone want to help as well on the player labelling, please feel free 🙏

**Reply by larsmaurath** [reply]
@DonsetPG  Of course no problem! We'll just go through it whenever we have time. Absolutely no pressure.

@alen1010 Great! I've downloaded it and saved it to my folder with the footage. Once we have a meaningful number I can upload everything together.

**Comment by larsmaurath** [comment]
**Example of Research on Corners**

Below a talk by Laurie Shaw on his research on corners that was based on roughly 1700 samples to show how this type of data could be used.

https://www.youtube.com/watch?v=yfPC1O_g-I8

**Comment by larsmaurath** [comment]
FYI, quick example of how the default narya weights perform when applying the full tracker to a typical corner scene. Homography estimation seems to work great as long as the penalty box is visible, but quite a few players remain untracked.


https://user-images.githubusercontent.com/11316760/111713114-4fdd4200-8825-11eb-8c18-9c68631fdedb.mp4

**Reply by DonsetPG** [reply]
haha yes def. not perfect, but could be a good starting point in terms of homography. 
A lot of work to do in terms of player tracking thought.

**Comment by kkoripl** [comment]
Hi @larsmaurath,
In the meantime of everything else, I'm trying to build my own tracking dataset and as I saw right now I have at least a few images with corner kicks, so I thought maybe you'd like to also look at them?
Don't know If you have particular way of labeling pictures (order maybe or when to actually label player) and in mine there are also 2 additional labels: GK and referee, but maybe it would be useful.
Let me know!

**Reply by larsmaurath** [reply]
Hi @kkoripl, in a thread above I shared some example images of how I label (https://github.com/DonsetPG/narya/discussions/27#discussioncomment-563138). I followed the Narya way of only labeling all humans and the ball.

Are you successful with differentiating between GKs and the referee with a trained model?
Happy to have a look at your data and I can check how easy it is to merge with the existing data.

**Reply by kkoripl** [reply]
Hi Lars, bad thing - I realized I've made a mistake to label balls, which are not on the pitch (lots of my images have i.e. balls around the pitch as there is no ball-boys), so first tests on my images for ball recognition are pretty bad. Would need to change it.
If you're still interested in the annotations, let me know.

About model, I've trained it on about 25% of my data, as I'm still (and would be probably pretty long...) in process of labeling, but... it's looking promising. I've pictures where GK's are seen almost samely good as players (but i think model learned in the way, that GK has a line behind him somewhere near waist, just my observation). With refs it's a little worse, but still promising. Linesmen are boxed well (but same thing as with GK's I think model learned, that linesman have line under his boots), when main refs (same class for me as linesman) are needed to be trained on the bigger data, cause yellow-black or red-black kit is boxed well, but complete black is terrible. That can be just my data, as I made parting to sets on random.

**Comment by larsmaurath** [comment]
**Overview of Labeling Tasks:**

| Batch   |      Images      | Event Data  | Homography Labels | Player Labels | Link |
|:--------|--------------|-------------|---------------------|--------------|------|
| 1           |  177                | larsmaurath |    alen1010 (done)    | opengoalapp (done) | [Link](https://www.dropbox.com/s/mlub6efhbj7wpfe/Batch_1_Frames.zip?dl=0)
| 2           |  191                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/u24px42toou4oog/Batch_2_Frames.zip?dl=0)
| 3           |  137                | larsmaurath |                                  | opengoalapp | [Link](https://www.dropbox.com/s/ah44g62x726tar3/Batch_3_Frames.zip?dl=0)
| 4           |  147                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/gwt2uvsd76bymzv/Batch_4_Frames.zip?dl=0)
| 5           |  200                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/8kshxb6szkrf9fa/Batch_5_Frames.zip?dl=0)
| 6           |  198                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/njav8jdilig7ssz/Batch_6_Frames.zip?dl=0)
| 7           |  200                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/yd4ugfanaabe17c/Batch_7_Frames.zip?dl=0)
| 8           |  191                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/4gqpor78jjpzy9r/Batch_8_Frames.zip?dl=0)
| 9           |  199                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/h9322obs9383lhe/Batch_9_Frames.zip?dl=0)
| 10           |  198                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/sqmlzhf2qe4nd21/Batch_10_Frames.zip?dl=0)
| 11           |  199                | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/pbdvb4emnb6l7d0/Batch_11_Frames.zip?dl=0)
| 12           |  92                 | larsmaurath |                                  |                       | [Link](https://www.dropbox.com/s/ky38t89au1wqn19/Batch_12_Frames.zip?dl=0)

Empty cells are unclaimed tasks. Feel free to reach out to start on one and I will update the table.

**Tools:**

Keypoint Labels: https://kkoripl.github.io/NaryaKeyPointsDatasetCreator/
Player Labels: https://github.com/larsmaurath/narya-label-creator

**Other tasks:**

1. Retrain Narya with additional corner footage
2. Automatically identify attacking/defending team based on bounding box colors
3. Set-up code that takes in mp4 footage and spits out tracking data

**Reply by opengoalapp** [reply]
@larsmaurath I'm currently making my way through Batch 1 - Player Labels. 

Really nice app btw!

**Reply by larsmaurath** [reply]
Cool! I am currently focusing on producing more footage/batches! The app is heavily based on https://github.com/rjtavares/BirdsPyView, so I really didn't do that much.

**Reply by opengoalapp** [reply]
By way of an update, I'm about half way through the batch 1 player labelling. Currently takes me about 2 mins per image so it is taking a while 😅. Hoping to ramp up a bit in the next few days.

**Reply by larsmaurath** [reply]
Cool, I am now also almost done with the footage. Then we can also start thinking about how to efficiently produce the data!

**Reply by opengoalapp** [reply]
@larsmaurath Batch 1 player annotations are complete. Link [here](https://www.dropbox.com/sh/j97uql4h3u1o2xo/AABo3c1TYoVClOADtO5vjs1Ga?dl=0)

Note image 34 is missing (I removed the jpeg also) - I messed that one up!

I'll move on to player labels for batch 3 next (as it is the smallest), the thinking being pitch marking recognition around the penalty area seems to be in a pretty good place already whereas player recognition is in more urgent need of more training data. Does that seem fair?

**Reply by larsmaurath** [reply]
Cool, great @opengoalapp ! Yes I would agree with your comment on the pitch markings.
 
Btw, I wouldn't say that we need to label all of the above footage. Hopefully we could already see improvements with 2 batches or so.

Another approach could be to find the footage the model has the most issues with. We would need to run the images through the default model and track how many players are found. We could then go from smallest to largest (this assumes that roughly the same number of players are present for each corner). For example we could label the worst 10% of each batch.

I am now done with all the footage and will upload the final batch after the final. I am also mostly done with 3) above (Set-up code that takes in mp4 footage and spits out tracking data). I'll try to set-up a Google Colab this week.

**Reply by larsmaurath** [reply]
Now uploaded the final batch that includes the corners from the final. I am making progress towards 2), albeit slowly. The idea I am using is to cluster the identified bounding boxes by their similarity/dominant colors (i.e. all boxes with a lot of red should be of the same team). If anybody has any other ideas or has made progress on their own, please let me know! Will post here once I get good results.

**Reply by larsmaurath** [reply]
Thanks @alen1010 I am basically doing the same thing! Good to know that I am on the right track

**Reply by opengoalapp** [reply]
Is it worth having some kind of occlusion detection either as it its own category or an additional flag when processing the detected bounding boxes? Probably a bit late for that as I'm guessing it would need a separate label when training? 

I'm just thinking of the cases when players are well hidden by a player from the opposite team (often in corner routines) and the team detection code would confidently assign the wrong team as a result. Whereas in these situations we would want weak assignment or some kind of "occluded" label so the tracking data generation code has help when trying to put together a realistic data stream.

**Reply by larsmaurath** [reply]
That's definitely an open question. I was hoping that this could be solved by post-processing. E.g. if we have 50 frames for one player and the player is occluded for 5, then potentially he would be flagged as Team A for 45 frames and Team B for 5. You could just label the player with Team A for all frames. But that's just a guess at this point how well it would work.

**Reply by DonsetPG** [reply]
I think the same method as the ReIdentification module could be helpful for this matter!

**Reply by larsmaurath** [reply]
I uploaded my half-done code to do the automatic team classification. Currently does not work well, but has the basic set-up if anybody wants to investigate too (i am off the next 2-3 weeks).
Also shows the full work-flow from video to animated tracking data. 

https://github.com/larsmaurath/narya-team-identification

![image](https://user-images.githubusercontent.com/11316760/122846187-528f7d80-d2d3-11eb-91ae-20132e7e86e6.png)
